enhancements/machine-config/install-time-support-for-image-mode-on-openshift.md

```diff
@@ -0,0 +1,343 @@
+
+#### Notes and Restrictions:
+
+- A pre-built custom OS container image is needed. Image build during bootstrap is not supported
+- The Containerfile used to build the image needs to be added to the MachineOSConfig CRs created at install time
+- Two boots will occur: Initial disk image boot with stock RHCOS image and a second boot when the MCD first boot service runs to boot into the custom OS container image
+- The registry where the pre-built image is stored will need to be accessible to the cluster in future as any future builds will be pushed there and pulled from there when rolling out onto the nodes
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+We currently don't have support for Hypershift with post-install Image Mode. Install time support for this will be added (if needed) once we have post-install support for it. This is currently not part of the implementation plan for this enhancement.
+
+#### Standalone Clusters
+
+Clusters in disconnected environments require special environment configurations such as a local package mirror, local image registry, etc.
+Post-install Image Mode currently works for disconnected environments and should work in a similar manner for install time
+
+#### Single-node Deployments or MicroShift
+
+Single-Node OpenShift (SNO) currently works fine with post-install Image Mode and should continue to work as-is for install time support as well.
+
+### API Extensions
+
+None
+
+### Implementation Details/Notes/Constraints
+
+#### Bootstrap Process Integration
+
+The bootstrap MCO will be enhanced to:
+
+- Parse MOSC configurations from the installation directory
+- Update the MCD first boot service to use the seeded pre-built custom OS container image
+
+#### Node Deployment Process
+
+1. Nodes initially boot with stock RHCOS OS
+2. MCD first boot service triggers `rpm-ostree rebase` or `bootc switch` to custom images
+3. Nodes boot into custom OS container images
+
+#### Post-install Integration
+
+1. A "fake" MachineOSBuild is created for the seeded pre-built container image
+2. This MOSB is set to successful state so the build controller is able to keep track of what needs to happen for any future updates or reverts
+3. The MOSC added at install time is used to maintain the customizations across updates
+
+#### MachineOSConfig (MOSC) Specifications
+
+MOSC YAML files for both `master` and `worker` MachineConfigPools will be placed in the installation directory and parsed by the bootstrap MCO. Here is an example:
+
+```yaml
+# mosc-worker.yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineOSConfig
+metadata:
+  name: worker
+  annotations:      
+    # Key annotation that triggers hybrid workflow                
+    machineconfiguration.openshift.io/pre-built-image: "registry.example.com/custom-rhcos-worker:latest@sha256:abc123def456..."
```

Comment:
following up from the question above, is this mandatory to be provided via the user? What happens if they provide a MOSC without this? Should we fail the install?

Reply:
Yes, this is mandatory and needs to be provided the user. I plan to add some validations in place to check for this and  fail if this is not provided.

Reply:
I guess it's fine to say "don't put that into the install manifests and do it purely as a day 2 operation" which I guess prevents any race conditions from our end